### ISSUE TITLE
[Un]-Bounded Array

### ISSUE THREAD

**Issue by bassoy** [BoostGSoC18/tensor#7]
- We need a `std::vector` and `std::array` conforming storage template classes in order to be able to provide it as a template parameter (define a type requirements for storage containers)
- Why do we need extra storage template classes like `boost::numeric::ublas::unbounded_array` and `boost::numeric::ublas::bounded_array` if `std::vector` and `std::array`exists?

**Comment by stefanseefeld** [comment]
The `Storage` concept is documented in https://www.boost.org/doc/libs/1_67_0/libs/numeric/ublas/doc/storage_concept.html. To cite from https://www.boost.org/doc/libs/1_67_0/libs/numeric/ublas/doc/unbounded_array.html:
"The unbounded array is similar to a std::vector in that in can grow in size beyond any fixed bound. However unbounded_array is aimed at optimal performance. Therefore unbounded_array does not model a Sequence like std::vector does. "

**Comment by bassoy** [comment]
Yes. However, I am not sure of what _optimal performance_ refers to and why unbounded array should not model a sequence container just because the resizing strategy might be different. Also there has been a discussion about this, see [CppCoreGuidelines](https://github.com/isocpp/CppCoreGuidelines/issues/493).

**Comment by stefanseefeld** [comment]
As far as I understand this is all about std::vector (and other sequences) initializing their elements, which we don't want / need as initialization is done by the tensor class itself.

**Comment by bassoy** [comment]
`std::vector` and other containers have a default constructor constructing an empty container. 
 `boost::numeric::ublas::tensor` has also a default constructor with the same behavior. So `boost::numeric::ublas::tensor<float> A;` does not allocate memory and does not initialize itself.
